pkg/common-controller/snapshot_controller.go

```diff
@@ -173,6 +171,19 @@ func (ctrl *csiSnapshotCommonController) syncSnapshot(snapshot *crdv1.VolumeSnap
 	if snapshot.ObjectMeta.DeletionTimestamp != nil && !ctrl.isPVCwithFinalizerInUseByCurrentSnapshot(snapshot) {
 		return ctrl.processSnapshotWithDeletionTimestamp(snapshot)
 	}
+
+	// TODO(xiangqian@): Putting this check in controller as webhook has not been implemented
+	//                   yet. Remove the source checking one issue #187 is resolved:
+	//                   https://github.com/kubernetes-csi/external-snapshotter/issues/187
+	klog.V(5).Infof("syncSnapshot[%s]: validate snapshot to make sure source has been correctly specified", utils.SnapshotKey(snapshot))
+	if (snapshot.Spec.Source.PersistentVolumeClaimName == nil && snapshot.Spec.Source.VolumeSnapshotContentName == nil) ||
+		(snapshot.Spec.Source.PersistentVolumeClaimName != nil && snapshot.Spec.Source.VolumeSnapshotContentName != nil) {
+		err := fmt.Errorf("Exactly one of PersistentVolumeClaimName and VolumeSnapshotContentName should be specified")
+		klog.Errorf("syncSnapshot[%s]: validation error, %s", utils.SnapshotKey(snapshot), err.Error())
+		ctrl.updateSnapshotErrorStatusWithEvent(snapshot, v1.EventTypeWarning, "SnapshotValidationError", err.Error())
+		return err
+	}
+
```

Comment:
We should check for content source in syncContent too just in case it is modified by mistake by user?

Reply:
Putting the check ONLY in syncContent is enough, any getXXXContent call should be verifying as syncSnapshot is the one conducting the binding.
I am open to have a check in syncContent and issue an event attach to the content if that helps. WDYT?